pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,346 @@
+			if all {
+				nsList, err := kc.Namespaces().List(labels.Everything())
+				checkErr(err)
+				for _, ns := range nsList.Items {
+					namespaces = append(namespaces, ns.Name)
+				}
+			} else {
+				namespaces = append(namespaces, defaultNamespace)
+			}
+
+			// Get all build configurations
+			buildConfigList := make([]buildapi.BuildConfig, 0)
+			for _, namespace := range namespaces {
+				cfgList, err := osc.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+				checkErr(err)
+				for _, cfg := range cfgList.Items {
+					buildConfigList = append(buildConfigList, cfg)
+				}
+			}
+
+			// Get all image repositories from build configurations
+			repos := getRepos(buildConfigList, defaultNamespace)
+
+			var tag string
+			if !all && len(args) != 0 {
+				var repo string
+				repo, tag, err = breakDown(args[0])
+				checkErr(err)
+
+				// Validate the specified image repository
+				if _, ok := repos[repo]; !ok {
+					checkErr(fmt.Errorf("invalid image repository: %s", repo))
+				}
+
+				flushMap(repos)
+				repos[repo] = defaultNamespace
+			} else {
+				// Make sure from now on that the --all flag is true
+				// since we are building dependency trees for every
+				// image repository available either in the current
+				// namespace or in every namespace
+				all = true
+			}
+
+			output := cmdutil.GetFlagString(cmd, "output")
+			for repo, namespace := range repos {
+				root := findRepoDeps(repo, namespace, tag, all, buildConfigList)
+
+				switch output {
+				case "json":
+					jsonDump, err := json.MarshalIndent(root, "", "\t")
+					checkErr(err)
+					fmt.Println(string(jsonDump))
+				case "dot":
+					g := dot.NewGraph()
+					// The DOT parser yells on names with dashes
+					graphName := strings.Replace(repo, "-", "_", -1)
+					g.SetName(graphName)
+					// Directed graph since we illustrate dependencies
+					g.SetDir(true)
+					// Explicitly allow multiple pairs of edges between
+					// the same pair of nodes
+					g.SetStrict(false)
+					g.AddAttr(graphName, "namespace", namespace)
+					fmt.Println(dotDump(root, g, graphName))
+				case "ast":
+					fmt.Println(root)
+				default:
+					usageError(cmd, "Wrong output format specified: %s", output)
+				}
+			}
+		},
+	}
+
+	cmd.Flags().Bool("all", false, "Build dependency trees for all image repositories")
+	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	return cmd
+}
+
+// getRepos iterates over a given set of build configurations
+// and extracts all the image repositories on which other
+// image repositories are based
+func getRepos(configs []buildapi.BuildConfig, defaultNamespace string) map[string]string {
+	avoidDuplicates := make(map[string]string)
+	for _, cfg := range configs {
+		for _, tr := range cfg.Triggers {
+			if tr.ImageChange != nil {
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					avoidDuplicates[tr.ImageChange.From.Name] = defaultNamespace
```

Comment:
Shouldn't the key to avoidDuplicates include the namespace? Otherwise you're going to only have one entry for the same name across namespaces